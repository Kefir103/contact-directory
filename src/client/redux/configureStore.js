import { createStore, applyMiddleware, compose } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/index';
import { rootReducer } from './reducers/rootReducer';
import thunk from 'redux-thunk';

let store;

export default function configureStore(initialState) {
    if (store) {
        return store;
    }

    store =
        process.env.NODE_ENV === 'production'
            ? createStore(rootReducer, initialState, compose(applyMiddleware(thunk)))
            : createStore(rootReducer, initialState, composeWithDevTools(applyMiddleware(thunk)));

    return store;
}
