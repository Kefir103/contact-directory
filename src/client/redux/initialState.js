export const initialState = {
    filter: {
        sortingMap: new Map([]),
        filterText: '',
        filterFields: [],
        currentPage: 1,
    },
    data: {
        currentElement: {
            elementInfo: undefined,
            elementIndex: -1,
        },
        elements: [],
        currentElements: [],
        filteredElements: undefined,
        inputContainer: {
            isAddButtonDisabled: false,
            isFormOpen: false,
            inputElement: {
                id: 0,
                firstName: '',
                lastName: '',
                email: '',
                phone: '',
            },
            validInputs: {
                id: false,
                firstName: false,
                lastName: false,
                email: false,
                phone: false,
            },
            addingStatus: {
                addingStatusText: '',
                isError: false,
            },
        },
    },
    appStatus: {
        isLoading: false,
        error: undefined,
    },
};
