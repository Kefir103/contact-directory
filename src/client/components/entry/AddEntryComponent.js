import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
    changeInputElementField,
    changeValidateInputs,
    changeIsAddButtonDisabled,
    changeIsInputFormOpen,
    removeInputData,
    setData,
    changeInputAddingStatus,
    resetValidInputs,
} from '../../redux/actions/dataActions';

const AddEntryComponent = (props) => {
    const areFieldsTrue = (object) => {
        for (const field in object) {
            if (!object[field]) {
                return false;
            }
        }
        return true;
    };

    const validateInput = (event, pattern) => {
        const regExp = new RegExp(pattern);
        const fieldName = event.target.name;

        const isValid = regExp.test(event.target.value);
        const nowValidInputs = { ...props.validInputs };
        nowValidInputs[fieldName] = isValid;

        props.actions.changeValidateInputs(fieldName, isValid);
        if (areFieldsTrue(nowValidInputs)) {
            props.actions.changeIsAddButtonDisabled(false);
        } else {
            props.actions.changeIsAddButtonDisabled(true);
        }

        return nowValidInputs[fieldName];
    };

    const handleInputIdChange = (event) => {
        validateInput(event, /^(0|[1-9][0-9]*)$/g);
        props.actions.changeInputElementField(event.target.name, Number(event.target.value));
    };

    const handleInputNameChange = (event) => {
        validateInput(event, /^[A-Za-z]+(\s[A-Z]*[a-z]*)*$/);
        props.actions.changeInputElementField(event.target.name, event.target.value);
    };

    const handleInputEmailChange = (event) => {
        validateInput(
            event,
            /^[A-Za-z0-9]+([_\-]?[A-Za-z0-9]+)*@[A-Za-z0-9]+([_\-]?[A-Za-z0-9]+)*(\.[_\-]?[A-Za-z0-9]+)*(\.[a-z]+)$/
        );
        props.actions.changeInputElementField(event.target.name, event.target.value);
    };

    const handleInputPhoneChange = (event) => {
        validateInput(event, /^\([0-9]{3}\)[0-9]{3}-[0-9]{4}$/);
        props.actions.changeInputElementField(event.target.name, event.target.value);
    };

    const handleAddButtonClick = (event) => {
        const isFormOpen = props.isFormOpen;
        if (!isFormOpen) {
            props.actions.changeIsInputFormOpen(true);
            props.actions.changeIsAddButtonDisabled(true);
        } else {
            if (
                props.elements.findIndex((element) => element.id === props.inputElement.id) !== -1
            ) {
                props.actions.changeInputAddingStatus('Элемент с таким id уже существует!', true);
            } else {
                props.actions.changeInputAddingStatus('Пользователь успешно добавлен!', false);
                props.actions.setData([props.inputElement, ...props.elements], props.sortingMap);
            }
            props.actions.changeIsAddButtonDisabled(false);
        }
    };

    const handleCloseFormButtonClick = (event) => {
        props.actions.changeIsInputFormOpen(false);
        props.actions.changeIsAddButtonDisabled(false);
        props.actions.removeInputData();
        props.actions.changeInputAddingStatus('', false);
        props.actions.resetValidInputs();
    };

    return (
        <div className={'add-entry-component'}>
            <button
                className={'add-button'}
                disabled={props.isAddButtonDisabled}
                onClick={handleAddButtonClick}>
                Добавить
            </button>
            <div
                className={'add-entry-component input-form'}
                style={props.isFormOpen ? { display: 'flex' } : { display: 'none' }}>
                <div className={'add-entry-component inputs'}>
                    <label className={'add-entry-component label-input'}>
                        Id
                        <input
                            type={'text'}
                            name={'id'}
                            placeholder={'Id'}
                            onChange={handleInputIdChange}
                            value={props.inputElement.id || ''}
                            autoComplete={'off'}
                            className={
                                props.validInputs.id || !props.inputElement.id ? '' : 'error-input'
                            }
                        />
                    </label>
                    <label className={'add-entry-component label-input'}>
                        First Name
                        <input
                            type={'text'}
                            name={'firstName'}
                            placeholder={'First Name'}
                            onChange={handleInputNameChange}
                            value={props.inputElement.firstName}
                            autoComplete={'off'}
                            className={
                                props.validInputs.firstName || !props.inputElement.firstName
                                    ? ''
                                    : 'error-input'
                            }
                        />
                    </label>
                    <label className={'add-entry-component label-input'}>
                        Last Name
                        <input
                            type={'text'}
                            name={'lastName'}
                            placeholder={'Last Name'}
                            onChange={handleInputNameChange}
                            value={props.inputElement.lastName}
                            autoComplete={'off'}
                            className={
                                props.validInputs.lastName || !props.inputElement.lastName
                                    ? ''
                                    : 'error-input'
                            }
                        />
                    </label>
                    <label className={'add-entry-component label-input'}>
                        Email
                        <input
                            type={'text'}
                            name={'email'}
                            placeholder={'Email'}
                            onChange={handleInputEmailChange}
                            value={props.inputElement.email}
                            autoComplete={'off'}
                            className={
                                props.validInputs.email || !props.inputElement.email
                                    ? ''
                                    : 'error-input'
                            }
                        />
                    </label>
                    <label className={'add-entry-component label-input'}>
                        Phone
                        <input
                            type={'text'}
                            name={'phone'}
                            placeholder={'Phone'}
                            onChange={handleInputPhoneChange}
                            value={props.inputElement.phone}
                            autoComplete={'off'}
                            className={
                                props.validInputs.phone || !props.inputElement.phone
                                    ? ''
                                    : 'error-input'
                            }
                        />
                    </label>
                </div>
                <p
                    className={
                        props.addingStatus.isError ? 'input-error-status' : 'input-fine-status'
                    }>
                    {props.addingStatus.addingStatusText}
                </p>
                <button className={'close-form-button'} onClick={handleCloseFormButtonClick}>
                    Закрыть окно
                </button>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        inputElement: state.data.inputContainer.inputElement,
        validInputs: state.data.inputContainer.validInputs,
        isFormOpen: state.data.inputContainer.isFormOpen,
        isAddButtonDisabled: state.data.inputContainer.isAddButtonDisabled,
        elements: state.data.elements,
        sortingMap: state.filter.sortingMap,
        addingStatus: state.data.inputContainer.addingStatus,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        actions: bindActionCreators(
            {
                changeInputElementField,
                changeValidateInputs,
                changeIsAddButtonDisabled,
                changeIsInputFormOpen,
                removeInputData,
                setData,
                changeInputAddingStatus,
                resetValidInputs,
            },
            dispatch
        ),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(AddEntryComponent);
