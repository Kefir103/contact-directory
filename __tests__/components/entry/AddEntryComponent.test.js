import React from 'react';
import configureMockStore from 'redux-mock-store';
import AddEntryComponent from '../../../src/client/components/entry/AddEntryComponent';
import { initialState } from '../../../src/client/redux/initialState';
import { shallow, render } from 'enzyme';

const mockStore = configureMockStore();
describe('<AddEntryComponent />', () => {
    let store;
    beforeEach(() => {
        store = mockStore({ ...initialState });
    });
    test('should render correctly with initialState', () => {
        const wrapper = shallow(<AddEntryComponent store={store} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
    test('should render correctly with isFormOpen', () => {
        const changedStore = mockStore({
            ...initialState,
            data: {
                ...initialState.data,
                inputContainer: {
                    ...initialState.data.inputContainer,
                    isFormOpen: true,
                },
            },
        });
        const wrapper = shallow(<AddEntryComponent store={changedStore} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
    test('should render correctly with isAddButtonDisabled and isFormOpen', () => {
        const changedStore = mockStore({
            ...initialState,
            data: {
                ...initialState.data,
                inputContainer: {
                    ...initialState.data.inputContainer,
                    isAddButtonDisabled: true,
                    isFormOpen: true,
                },
            },
        });
        const wrapper = shallow(<AddEntryComponent store={changedStore} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
    test('should render correctly with filled but not validated inputs', () => {
        const changedStore = mockStore({
            ...initialState,
            data: {
                ...initialState.data,
                inputContainer: {
                    ...initialState.data.inputContainer,
                    isAddButtonDisabled: true,
                    isFormOpen: true,
                    inputElement: {
                        id: 1,
                        firstName: 'firstName',
                        lastName: 'lastName',
                        email: 'email',
                        phone: 'phone',
                    },
                },
            },
        });
        const wrapper = shallow(<AddEntryComponent store={changedStore} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
    test('should render correctly with good addingStatus', () => {
        const changedStore = mockStore({
            ...initialState,
            data: {
                ...initialState.data,
                inputContainer: {
                    ...initialState.data.inputContainer,
                    addingStatus: {
                        addingStatusText: 'addingStatusText',
                        isError: false,
                    },
                },
            },
        });
        const wrapper = shallow(<AddEntryComponent store={changedStore} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
    test('should render correctly with error addingStatus', () => {
        const changedStore = mockStore({
            ...initialState,
            data: {
                ...initialState.data,
                inputContainer: {
                    ...initialState.data.inputContainer,
                    addingStatus: {
                        addingStatusText: 'addingStatusText',
                        isError: true,
                    },
                },
            },
        });
        const wrapper = shallow(<AddEntryComponent store={changedStore} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
});
