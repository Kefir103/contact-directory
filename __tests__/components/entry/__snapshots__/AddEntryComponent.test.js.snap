// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddEntryComponent /> should render correctly with error addingStatus 1`] = `
<div
  class="add-entry-component"
>
  <button
    class="add-button"
  >
    Добавить
  </button>
  <div
    class="add-entry-component input-form"
    style="display:none"
  >
    <div
      class="add-entry-component inputs"
    >
      <label
        class="add-entry-component label-input"
      >
        Id
        <input
          autocomplete="off"
          class=""
          name="id"
          placeholder="Id"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        First Name
        <input
          autocomplete="off"
          class=""
          name="firstName"
          placeholder="First Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Last Name
        <input
          autocomplete="off"
          class=""
          name="lastName"
          placeholder="Last Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Email
        <input
          autocomplete="off"
          class=""
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Phone
        <input
          autocomplete="off"
          class=""
          name="phone"
          placeholder="Phone"
          type="text"
          value=""
        />
      </label>
    </div>
    <p
      class="input-error-status"
    >
      addingStatusText
    </p>
    <button
      class="close-form-button"
    >
      Закрыть окно
    </button>
  </div>
</div>
`;

exports[`<AddEntryComponent /> should render correctly with filled but not validated inputs 1`] = `
<div
  class="add-entry-component"
>
  <button
    class="add-button"
    disabled=""
  >
    Добавить
  </button>
  <div
    class="add-entry-component input-form"
    style="display:flex"
  >
    <div
      class="add-entry-component inputs"
    >
      <label
        class="add-entry-component label-input"
      >
        Id
        <input
          autocomplete="off"
          class="error-input"
          name="id"
          placeholder="Id"
          type="text"
          value="1"
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        First Name
        <input
          autocomplete="off"
          class="error-input"
          name="firstName"
          placeholder="First Name"
          type="text"
          value="firstName"
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Last Name
        <input
          autocomplete="off"
          class="error-input"
          name="lastName"
          placeholder="Last Name"
          type="text"
          value="lastName"
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Email
        <input
          autocomplete="off"
          class="error-input"
          name="email"
          placeholder="Email"
          type="text"
          value="email"
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Phone
        <input
          autocomplete="off"
          class="error-input"
          name="phone"
          placeholder="Phone"
          type="text"
          value="phone"
        />
      </label>
    </div>
    <p
      class="input-fine-status"
    />
    <button
      class="close-form-button"
    >
      Закрыть окно
    </button>
  </div>
</div>
`;

exports[`<AddEntryComponent /> should render correctly with good addingStatus 1`] = `
<div
  class="add-entry-component"
>
  <button
    class="add-button"
  >
    Добавить
  </button>
  <div
    class="add-entry-component input-form"
    style="display:none"
  >
    <div
      class="add-entry-component inputs"
    >
      <label
        class="add-entry-component label-input"
      >
        Id
        <input
          autocomplete="off"
          class=""
          name="id"
          placeholder="Id"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        First Name
        <input
          autocomplete="off"
          class=""
          name="firstName"
          placeholder="First Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Last Name
        <input
          autocomplete="off"
          class=""
          name="lastName"
          placeholder="Last Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Email
        <input
          autocomplete="off"
          class=""
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Phone
        <input
          autocomplete="off"
          class=""
          name="phone"
          placeholder="Phone"
          type="text"
          value=""
        />
      </label>
    </div>
    <p
      class="input-fine-status"
    >
      addingStatusText
    </p>
    <button
      class="close-form-button"
    >
      Закрыть окно
    </button>
  </div>
</div>
`;

exports[`<AddEntryComponent /> should render correctly with initialState 1`] = `
<div
  class="add-entry-component"
>
  <button
    class="add-button"
  >
    Добавить
  </button>
  <div
    class="add-entry-component input-form"
    style="display:none"
  >
    <div
      class="add-entry-component inputs"
    >
      <label
        class="add-entry-component label-input"
      >
        Id
        <input
          autocomplete="off"
          class=""
          name="id"
          placeholder="Id"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        First Name
        <input
          autocomplete="off"
          class=""
          name="firstName"
          placeholder="First Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Last Name
        <input
          autocomplete="off"
          class=""
          name="lastName"
          placeholder="Last Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Email
        <input
          autocomplete="off"
          class=""
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Phone
        <input
          autocomplete="off"
          class=""
          name="phone"
          placeholder="Phone"
          type="text"
          value=""
        />
      </label>
    </div>
    <p
      class="input-fine-status"
    />
    <button
      class="close-form-button"
    >
      Закрыть окно
    </button>
  </div>
</div>
`;

exports[`<AddEntryComponent /> should render correctly with isAddButtonDisabled and isFormOpen 1`] = `
<div
  class="add-entry-component"
>
  <button
    class="add-button"
    disabled=""
  >
    Добавить
  </button>
  <div
    class="add-entry-component input-form"
    style="display:flex"
  >
    <div
      class="add-entry-component inputs"
    >
      <label
        class="add-entry-component label-input"
      >
        Id
        <input
          autocomplete="off"
          class=""
          name="id"
          placeholder="Id"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        First Name
        <input
          autocomplete="off"
          class=""
          name="firstName"
          placeholder="First Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Last Name
        <input
          autocomplete="off"
          class=""
          name="lastName"
          placeholder="Last Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Email
        <input
          autocomplete="off"
          class=""
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Phone
        <input
          autocomplete="off"
          class=""
          name="phone"
          placeholder="Phone"
          type="text"
          value=""
        />
      </label>
    </div>
    <p
      class="input-fine-status"
    />
    <button
      class="close-form-button"
    >
      Закрыть окно
    </button>
  </div>
</div>
`;

exports[`<AddEntryComponent /> should render correctly with isFormOpen 1`] = `
<div
  class="add-entry-component"
>
  <button
    class="add-button"
  >
    Добавить
  </button>
  <div
    class="add-entry-component input-form"
    style="display:flex"
  >
    <div
      class="add-entry-component inputs"
    >
      <label
        class="add-entry-component label-input"
      >
        Id
        <input
          autocomplete="off"
          class=""
          name="id"
          placeholder="Id"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        First Name
        <input
          autocomplete="off"
          class=""
          name="firstName"
          placeholder="First Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Last Name
        <input
          autocomplete="off"
          class=""
          name="lastName"
          placeholder="Last Name"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Email
        <input
          autocomplete="off"
          class=""
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </label>
      <label
        class="add-entry-component label-input"
      >
        Phone
        <input
          autocomplete="off"
          class=""
          name="phone"
          placeholder="Phone"
          type="text"
          value=""
        />
      </label>
    </div>
    <p
      class="input-fine-status"
    />
    <button
      class="close-form-button"
    >
      Закрыть окно
    </button>
  </div>
</div>
`;
